### PR TITLE
Add documentation warnings for weak algorithms

### PR DESCRIPTION
**Pull request by hanno-arm**
__Summary:__ This PR adds warnings for some weak algorithms ~~and deprecates a few of them~~:
1. It adds documentation warnings regarding the use of MD2, MD4, MD5, DES, ARC4 and SHA-1:
a. Any public API function gets an appropriate Doxygen `\warning`.
b. `config.h` is enhanced by `\warning`'s for the respective configuration option.
c. The documentation of identifier types relating to these algorithms (like `mbedtls_md_type_t` containng `MBEDTLS_MD_MD2`) are enhanced by an appropriate Doxygen `\warning`.
~~2. It deprecates MD2, MD4 and ARC4: 
a. A single deprecation warning is emitted for `md2.h`, `md4.h` and `arc4.h`, and all public functions related to these algorithms get a Doxygen `\deprecated` flag.
b. Identifiers related to these algorithms are deprecated.
c. Configuration options related to these algorithms get a deprecation note in addition to the warnings mentioned above.~~

~~This leads to the following behavior: 1. Compilation without `DEPRECATED_WARNING` succeeds without any warnings. In particular, the default configuration builds without warnings.
2. Compilation with `DEPRECATED_WARNING` with either `MD2_C`, `MD4_C` or `ARC4_C` enabled results in numerous warnings from includes, ID uses, etc. 
3. As a consequence of 2., it is not possible to build in modes like `ASanDbg` treating warnings as errors, when `DEPRECATED_WARNING` is set. While not optimal, I don't consider this critical. If you want to check where you use deprecated features, make a normal build and inspect warnings, and if you want to test your code in debug mode, disable `DEPRECATED_WARNING` again and build.~~

A small change to the test-code generating script `generate_code.pl` had to be made in order to omit mentioning identifiers like `MBEDTLS_MD2_MD` in case their respective modules were disabled; previously, this lead to unavoidable deprecation warnings even if no deprecated module was enabled.

### CODE REVIEW

File: include/mbedtls/sha1.h
Comment (andresag01) [inline include/mbedtls/sha1.h]: Please note that the module name is written `SHA-1` in the other documentation

**Comment by hanno-arm** [comment]
Thanks @andresag01 for the quick feedback. I'll add comments at the top of the files and for the context structures, too. Regarding the wording, what about "It is recommended to use a strong cipher/message digest instead." instead?

**Comment by andresag01** [comment]
@hanno-arm: Yep, I am happy with something like that. I am definitely agains saying something like: "use AES instead of ARC4" because I do not think it is our place to give security advice. Also whatever you recommend might be out-of-date soon (as unlikely as it might be...)

**Comment by hanno-arm** [comment]
@andresag01 I have updated the PR according to your comments.

File: include/mbedtls/config.h
Comment (andresag01) [inline include/mbedtls/config.h]: Should we add a similar warning to `MBEDTLS_TLS_DEFAULT_ALLOW_SHA1_IN_KEY_EXCHANGE`?

File: include/mbedtls/md2.h
Comment (andresag01) [inline include/mbedtls/md2.h]: I think this warning is missing on `mbedtls_md2_finish()`

File: include/mbedtls/arc4.h
Comment (andresag01) [inline include/mbedtls/arc4.h]: Should we add a similar warning to the `mbedtls_arc4_self_test()`?

Comment (hanno-arm) [reply include/mbedtls/arc4.h]: I omitted these deliberately, but I don't really care - it's perhaps more uniform to add it there, too.

File: include/mbedtls/md2.h
Comment (andresag01) [reply include/mbedtls/md2.h]: Please note that this is not the only file that is missing warnings in some functions...

**Comment by hanno-arm** [comment]
Thanks and apologies @andresag01, that was careless. I updated the PR, including two more warnings regarding the configuration options guarding the use of SHA-1 for certificate signing and within key exchanges.

**Comment by hanno-arm** [comment]
@andresag01 Regarding DES, the main goal of this PR is so serve the internal requirements, so I'd prefer to stick to these and leave a further deprecation to a potential separate PR. 

Regarding the `mbedtls_md_type_t` enum, good point, that's another part of the code that the user interfaces with, so we should add the warning there, too.

**Comment by hanno-arm** [comment]
@andresag01 Updated.

**Comment by hanno-arm** [comment]
@andresag01 The `mbedtls_*_process` functions are completely undocumented and internal only judging from the respective header files, so from that perspective I'd be inclined to omit the warnings for them in the headers. That said, though, I find it surprising that despite the lack of any documentation, there's a configuration option in `config.h` allowing the user to overwrite them. It's not very consistent. 

For now, I've added the warnings to the `MBEDTLS_XXX_ALT` config options as you requested, but kept the rest as it is.

**Comment by andresag01** [comment]
@hanno-arm: You mean is not consistent because there is no documentation for the function? I think the documentation was not omitted because it is supposed to be an internal function only...

**Comment by hanno-arm** [comment]
@andresag01 I don't understand your last sentence. 
The functions are meant to be internal only, hence undocumented - so far, so good. But then it doesn't make sense to expose them in `config.h`, allowing the user to overwrite them - looking at the documentation only, the user has no idea what functionality he's supposed to provide.

**Comment by andresag01** [comment]
@hanno-arm: Its internal only in the sense that is only exposed for the purposes of alternative implementations. So if you want to replace `mbedtls_*_process()` then just define the macro and provide an implementation for the function.

**Comment by hanno-arm** [comment]
@andresag01 How can I re-implement a function if I have no information on the functionality it is supposed to provide? That's odd. 

Anyways, I've added a warning to the config options for `XXX_PROCESS_ALT`.

**Comment by andresag01** [comment]
@hanno-arm: btw, are we going to backport these changes?

**Comment by hanno-arm** [comment]
@andresag01 I don't know about internal requirements, but I think as these are warnings they should be backported.

File: include/mbedtls/arc4.h
Comment (sbutcher-arm) [inline include/mbedtls/arc4.h]: The problem with these warnings are it begs the question ‘what is a strong cipher’ or ‘strong message digest’? That is a question we want don't want to have to answer in support or in the forums, as users need to choose for themselves what suits them best. (And there's no end of bodies such as NIST or ENISA who'll recommend something).

Therefore I would prefer the wording:


> ARC4 is considered a weak cipher and its use constitutes a security risk. It is recommended alternative ciphers should be considered instead.

Comment (hanno-arm) [reply include/mbedtls/arc4.h]: @sbutcher-arm Thanks for that suggestion, I'll change the wordings accordingly.

Comment (hanno-arm) [reply include/mbedtls/arc4.h]: Done.

**Comment by gilles-peskine-arm** [comment]
I haven't checked in detail how we're going about deprecating these algorithms. But I'm surprised that DES isn't included. It isn't as badly broken as RC4 (if you're using 3DES), and it's the least bad ciphersuite to talk to Windows XP, but it is broken enough that we should be warning people away from it.

**Comment by gilles-peskine-arm** [comment]
Please ensure that deprecated functionality has a doxygen comment with the `\deprecated` macro (if it's activated by `config.h`). In the doxygen-generated documentation, `deprecated.html` is a convenient dashboard to review deprecated functionality — if we keep it accurate.

**Comment by hanno-arm** [comment]
@gilles-peskine-arm Yes, so far this PR hasn't been updated with any deprecation, but I'll use `\deprecated` for that. Thanks for mentioning the deprecation-dashboard, I didn't know about that.

**Comment by hanno-arm** [comment]
@sbutcher-arm @yanesca @andresag01 @mpg @gilles-peskine-arm  I have added commits deprecating MD2 and ARC4. I didn't do the other algorithms yet because I'd like to await your feedback in order to avoid redundant work if it's not what we want. 

Quick summary:
1. Add single deprecation warning to respective header files (`md2.h` and `arc4.h`)
2. Deprecate global identifiers, e.g. `MBEDTLS_MD_MD2` or ciphersuite ID's involving ARC4.
3. Make multiple includes of `md2.h` and `arc4.h` conditional on `MD2_C` resp. `ARC4_C` being set, to avoid deprecation warnings if the latter are not set.
4. Minor improvements, mostly style.

This leads to the following behavior:
1. Compilation without `DEPRECATED_WARNING` continues without any warnings. In particular, the default configuration builds without warnings.
2. Compilation with `DEPRECATED_WARNING` with either `MD2_C` or `ARC4_C` enabled results in numerous warnings from includes, ID uses, etc. As far as I understood, we tolerate that.
3. As a consequence of 2., it is not possible to build in modes like `ASanDbg` treating warnings as errors, when `DEPRECATED_WARNING` is set. While not optimal, I don't consider this critical. If you want to check where you use deprecated features, make a normal build and inspect warnings, and if you want to test your code in debug mode, disable `DEPRECATED_WARNING` again and build.
4. Compilation with `DEPRECATED_WARNING` enabled and `MD2_C` and `ARC4_C` both disabled currently still results in 3 warnings originating from automatically generated test-code which extracts identifiers like `MBEDTLS_MD2_MD` from the test files and unconditionally generates code for it which is not guarded by the corresponding `#if defined(MBEDTLS_MD2_C) ... #endif`. If we don't like it, we could investigate how to adapt the test-code generation scripts accordingly.

Happy about feedback as always.

**Comment by hanno-arm** [comment]
Removing `needs_work` temporarily while awaiting feedback before embarking on the other algorithms.

**Comment by gilles-peskine-arm** [comment]
@hanno-arm MD4 was broken the year SSLv2 came out and never made it into any ciphersuite. So please make it as deprecated as MD2. RC4 is less deprecated than MD4 (though I do support marking it as deprecated in mbedtls): the final nail in RC4's coffin was relatively recent, and I think it's still possible to use it while keeping ahead of the existing attacks (not that I'd recommend doing it, because there's a serious risk of falling to next year's incremental improvement to the attacks).

**Comment by hanno-arm** [comment]
@gilles-peskine-arm Please note that this is only incorporating MD2 and ARC4 to reach consensus before doing the other algorithms.

**Comment by andresag01** [comment]
@hanno-arm: What's the plan then? My current understanding is that we are adding doxygen warnings for MD2, MD4, MD5, SHA1, ARC4 and DES. However, we are only deprecating MD2, MD4 and ARC4. Is this correct? Also, I have a couple of comments:
* If ARC4 is being deprecated, why should we still have it enabled by default in the config.h?
* Should we consider enabling MBEDTLS_DEPRECATED_WARNING by default in the config.h?
* I think that we should look fixing the warnings in the deault config when all the deprecated modules are disabled, e.g. MD2, MD4 and ARC4.

**Comment by hanno-arm** [comment]
@andresag01 That's in accordance with my understanding. The current state of the PR adds warnings about MD2, MD4, MD5, SHA-1 and ARC4, so DES still needs to be added (no specific reason to postpone it, it just hasn't been done). Deprecation is exemplified in MD2 and ARC4 and pending your and the other's feedback to see if it's what we want.

Concerning your questions:
1. My feeling was that this would probably wait until the next API breaking release; in any case, it's not my decision, so I left it enabled so far.
2. Again, I'm open for that, but in this case we should disable ARC4 by default, otherwise we won't build without warning, resp. not at all in `ASanDbg`.
3. Yes, as mentioned, currently the only warnings we get are from deficiency in the code generation which doesn't guard code for string-to-identifier conversion by the configuration flags guarding the tests in which the respective identifiers were found. Concretely, there are tests containing `MBEDTLS_MD2_MD` in their test input and which depend on `MBEDTLS_MD2_C`, but the test-code generation doesn't guard the translation of the string `"MBEDTLS_MD2_MD"` to the identifier `MBEDTLS_MD2_MD` by `MBEDTLS_MD2_C`.

File: include/mbedtls/config.h
Comment (yanesca) [inline include/mbedtls/config.h]: The meaning of exclamation marks are pretty straightforward for us, but they are easy to overlook and might leave a sense of doubt in the reader. I suggest we clarify this a little bit by writing "Requires to disable the following macros:"

File: include/mbedtls/config.h
Comment (yanesca) [inline include/mbedtls/config.h]: I think we could emphasise the difference a little bit more, by saying "and can be used in TLS 1.2 ..."

Comment (hanno-arm) [reply include/mbedtls/config.h]: Done.

File: include/mbedtls/config.h
Comment (hanno-arm) [reply include/mbedtls/config.h]: Done.

**Comment by hanno-arm** [comment]
@yanesca There was a bug in the test code generation incaught by the CI for unknown reasons. Could you check and potentially re-approve?

File: include/mbedtls/arc4.h
Comment (gilles-peskine-arm) [inline include/mbedtls/arc4.h]: The warning should mention that we're planning to remove it (in the same terms as the documentation). It's important practical information at this point: even developers who know/think that their use case is safe should feel concerned.

Comment (gilles-peskine-arm) [inline include/mbedtls/arc4.h]: Having multiple `#warning` directives that shoukd read as a single paragraph is not very nice. GCC, at least, shows each directive separately. Please use a single directive (single line, or multiline string literal, or backslash continuations, whatever).

File: include/mbedtls/cipher.h
Comment (gilles-peskine-arm) [inline include/mbedtls/cipher.h]: With gcc 5.4.0 (Ubuntu 16.04 package), I just turned on `MBEDTLS_DEPRECATED_WARNING` in `config.h` and the compilation broke:

```  CC    ccm.c
In file included from ../include/mbedtls/ccm.h:26:0,
                 from ccm.c:39:
../include/mbedtls/cipher.h:70:28: error: expected ‘,’ or ‘}’ before ‘__attribute__’
 #define MBEDTLS_DEPRECATED __attribute__((deprecated))
                            ^
../include/mbedtls/cipher.h:93:28: note: in expansion of macro ‘MBEDTLS_DEPRECATED’
     MBEDTLS_CIPHER_ID_ARC4 MBEDTLS_DEPRECATED,
                            ^
../include/mbedtls/cipher.h:70:28: error: expected ‘,’ or ‘}’ before ‘__attribute__’
 #define MBEDTLS_DEPRECATED __attribute__((deprecated))
                            ^
../include/mbedtls/cipher.h:149:29: note: in expansion of macro ‘MBEDTLS_DEPRECATED’
     MBEDTLS_CIPHER_ARC4_128 MBEDTLS_DEPRECATED,
                             ^
Makefile:161: recipe for target 'ccm.o' failed
make: *** [ccm.o] Error 1
```

GCC supports `__attribute__((deprecated))` for functions, variables and types but not for enum constants.

As a workaround I suggest one of the solutions we've discussed before:

```
typedef enum { ... MBEDTLS_CIPHER_ID_ARC4_DEPRECATED ... } mbedtls_cipher_id_t;
typedef MBEDTLS_DEPRECATED mbedtls_deprecated_cipher_id_t mbedtls_cipher_id_t;
#define MBEDTLS_CIPHER_ID_ARC4 ( (mbedtls_deprecated_cipher_id_t) MBEDTLS_CIPHER_ID_ARC4_DEPRECATED )
```

File: include/mbedtls/ssl_ciphersuites.h
Comment (gilles-peskine-arm) [inline include/mbedtls/ssl_ciphersuites.h]: Please keep `typedef` in column 0 (and the type being defined last before the semicolon), it helps readers.

File: include/mbedtls/ssl_ciphersuites.h
Comment (gilles-peskine-arm) [inline include/mbedtls/ssl_ciphersuites.h]: I'd prefer these definitions to be in a single line for grepping. Unlike regular lines of codes, series of simple constant definitions like this don't suffer much if some lines are wrapped in casual reading. But when grepping, it's a pain to get the half you were searching and not the half you were trying to look up.

File: include/mbedtls/cipher.h
Comment (hanno-arm) [reply include/mbedtls/cipher.h]: @gilles-peskine-arm I am puzzled, I was sure I had tested this. Will look into it.

File: include/mbedtls/arc4.h
Comment (hanno-arm) [reply include/mbedtls/arc4.h]: Done.

Comment (hanno-arm) [reply include/mbedtls/arc4.h]: Done.

File: include/mbedtls/ssl_ciphersuites.h
Comment (hanno-arm) [reply include/mbedtls/ssl_ciphersuites.h]: Done.

File: include/mbedtls/ssl_ciphersuites.h
Comment (hanno-arm) [reply include/mbedtls/ssl_ciphersuites.h]: Done.

**Comment by hanno-arm** [comment]
@gilles-peskine-arm Thanks for your feedback. I've addressed your comments.

File: include/mbedtls/cipher.h
Comment (gilles-peskine-arm) [inline include/mbedtls/cipher.h]: It would be nice to make identifiers such as this one appear in the "Deprecated List" page of the Doxygen documentation. You can do that by adding a comment like
```
/** \deprecated ARC4 is deprecated and will likely be removed in a future version of the library. */
```
just before the `#define`.

**Comment by gilles-peskine-arm** [comment]
Note: CI failed only due to a failure of the timing selftest (known issue).

**Comment by hanno-arm** [comment]
@yanesca There have been some changes since your last approval - would you mind giving those another look?

**Comment by yanesca** [comment]
Looks good to me!

**Comment by hanno-arm** [comment]
Issue #1156 is fixed by this PR, too, and should be closed alongside it.

**Comment by hanno-arm** [comment]
Decision has been taken to remove all deprecations.

**Comment by hanno-arm** [comment]
@yanesca @gilles-peskine-arm As requested I have reverted all deprecations. Please have a look and potentially re-approve.

If you're ok with it, I strongly suggest to then clean up the history, keeping a copy of the present state so that you can verify there's zero diff after the rebase.

**Review by yanesca — DISMISSED** [review]

**Comment by hanno-arm** [comment]
For reference, the last commit is 493f43b2, I preserve it temporarily on a local branch for comparison.

**Comment by hanno-arm** [comment]
@yanesca @gilles-peskine-arm Dropped all `commit-revert` pairs from the PR. The diff between the current head 85f6c7d and 493f43b is empty.

**Review by simonbutcher — DISMISSED** [review]
Approved by me, but slightly nervous about merge conflicts.

**Comment by gilles-peskine-arm** [comment]
Merge conflicts:

* Deprecated hash functions — the new `_ret` hash functions need to be indicated as deprecated as well.
* Probably a few more with the upcoming documentation fixes.
* The conflict in the test suite is trivial.

I strongly recommend merging `development` into this PR and reviewing the merge.

**Comment by Patater** [comment]
I'm noticing multiple test failures with this PR, even before merging it into the development branch. The failures will need to be resolved before this PR can move forward.

Please test with `all.sh`.

**Comment by gilles-peskine-arm** [comment]
@Patater Here, all I'm seeing is failures related to old things that have been fixed since this PR was made. Please run tests **after merging** with current development.

**Comment by sbutcher-arm** [comment]
Hi @hanno-arm,

@Patater has submitted a new PR #1336, to include just the documentation fixes, to simplify merging this PR. I don't want to lose the other changes in this PR, so could you please adapt this PR or close it and submit a new PR that contains the other fixes in this PR which won't otherwise get included?

Thanks

**Comment by sbutcher-arm** [comment]
@hanno-arm - Can you give a status update on this? Thanks.

**Comment by hanno-arm** [comment]
As discussed with @sbutcher-arm, this PR is in hold for the moment. 

__Internal Reference:__ IOTSSL-2379

**Comment by daverodgman** [comment]
In 3.x, these algorithms are either removed, or have warnings.

**Pull request closed without merging**
